//===----------------------------------------------------------------------===//
//
// This source file is part of the Teco open source project.
//
// Copyright (c) 2022 the Teco project authors
// Licensed under Apache License v2.0
//
// See LICENSE.txt for license information
//
// SPDX-License-Identifier: Apache-2.0
//
//===----------------------------------------------------------------------===//

// THIS FILE IS AUTOMATICALLY GENERATED by TecoServiceGenerator.
// DO NOT EDIT.

extension TCApigatewayError {
    public struct InternalError: TCErrorType {
        enum Code: String {
            case apigwException = "InternalError.ApigwException"
            case cauthException = "InternalError.CauthException"
            case clbException = "InternalError.ClbException"
            case ossException = "InternalError.OssException"
            case scfException = "InternalError.ScfException"
            case tsfException = "InternalError.TsfException"
            case vpcException = "InternalError.VpcException"
            case other = "InternalError"
        }
        
        private let error: Code
        
        public let context: TCErrorContext?
        
        public var errorCode: String {
            self.error.rawValue
        }
        
        /// Initializer used by ``TCClient`` to match an error of this type.
        ///
        /// You should not use this initializer directly as there are no public initializers for ``TCErrorContext``.
        public init ?(errorCode: String, context: TCErrorContext) {
            guard let error = Code(rawValue: errorCode) else {
                return nil
            }
            self.error = error
            self.context = context
        }
        
        internal init (_ error: Code, context: TCErrorContext? = nil) {
            self.error = error
            self.context = context
        }
        
        /// API网关内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var apigwException: InternalError {
            InternalError(.apigwException)
        }
        
        /// CAuth内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var cauthException: InternalError {
            InternalError(.cauthException)
        }
        
        /// CLB内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var clbException: InternalError {
            InternalError(.clbException)
        }
        
        /// oss内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var ossException: InternalError {
            InternalError(.ossException)
        }
        
        /// SCF内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var scfException: InternalError {
            InternalError(.scfException)
        }
        
        /// TSF内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var tsfException: InternalError {
            InternalError(.tsfException)
        }
        
        /// vpc内部请求错误，请稍后重试。若无法解决，请联系智能客服或提交工单。
        public static var vpcException: InternalError {
            InternalError(.vpcException)
        }
        
        /// 内部错误。
        public static var other: InternalError {
            InternalError(.other)
        }
    }
}

extension TCApigatewayError.InternalError: Equatable {
    public static func == (lhs: TCApigatewayError.InternalError, rhs: TCApigatewayError.InternalError) -> Bool {
        lhs.error == rhs.error
    }
}

extension TCApigatewayError.InternalError: CustomStringConvertible {
    public var description: String {
        return "\(self.error.rawValue): \(message ?? "")"
    }
}

extension TCApigatewayError.InternalError {
    /// - Returns: ``TCApigatewayError`` that holds the same error and context.
    public func toApigatewayError() -> TCApigatewayError {
        guard let code = TCApigatewayError.Code(rawValue: self.error.rawValue) else {
            fatalError("Unexpected internal conversion error!\nPlease file a bug at https://github.com/teco-project/teco to help address the problem.")
        }
        return TCApigatewayError(code, context: self.context)
    }
}

extension TCApigatewayError.InternalError {
    /// - Returns: ``TCCommonError`` that holds the same error and context.
    public func toCommonError() -> TCCommonError? {
        if let context = self.context, let error = TCCommonError(errorCode: self.error.rawValue, context: context) {
            return error
        }
        return nil
    }
}
